src/function/cast/variant/to_variant.cpp

```diff
@@ -190,4 +191,214 @@ BoundCastInfo DefaultCasts::ImplicitToVariantCast(BindCastInput &input, const Lo
+					if (key_is_used[child_idx]) {
+						auto &struct_child = struct_children[child_idx];
+						string_t struct_child_str(struct_child.first.c_str(),
+						                          NumericCast<uint32_t>(struct_child.first.size()));
+						dictionary_indices.push_back(result.GetOrCreateIndex(struct_child_str));
+					} else {
+						dictionary_indices.push_back(static_cast<uint32_t>(-1));
+					}
+				}
+			}
+
+			//! children
+			if (WRITE_DATA) {
+				auto &children_list_entry = variant.children_data[result_index];
+				auto &keys_list_entry = variant.keys_data[result_index];
+
+				idx_t children_index = children_list_entry.offset + children_offset_data[result_index];
+				idx_t keys_offset = keys_list_entry.offset + keys_offset_data[result_index];
+
+				idx_t non_null_idx = 0; // Keep track of non-NULL children because this is a sparse conversion
+				for (idx_t child_idx = 0; child_idx < children.size(); child_idx++) {
+					const auto &child_format = child_source_data[child_idx].source_format;
+					if (!child_format.validity.RowIsValid(child_format.sel->get_index(i))) {
+						continue; // Sparse conversion: skip NULL
+					}
+					variant.keys_index_data[children_index + non_null_idx] =
+					    NumericCast<uint32_t>(keys_offset_data[result_index] + non_null_idx);
+					result.keys_selvec.set_index(keys_offset + non_null_idx, dictionary_indices[child_idx]);
+					non_null_idx++;
+				}
+				//! Map from index of the child to the children.values_index of the parent
+				//! NOTE: this maps to the first index, below we are forwarding this for each child Vector we process.
+				sel.children_selection.set_index(sel.count, children_index);
+			}
+			sel.non_null_selection.set_index(sel.count, source.GetMappedIndex(i));
+			sel.new_selection.set_index(sel.count, result_index);
+			keys_offset_data[result_index] += non_null_child_count;
+			children_offset_data[result_index] += non_null_child_count;
+			sel.count++;
+		} else if (!IGNORE_NULLS) {
+			HandleVariantNull<WRITE_DATA>(result, result_index, values_offset_data, blob_offset, values_index_selvec, i,
+			                              is_root);
+		}
+	}
+
+	for (idx_t child_idx = 0; child_idx < children.size(); child_idx++) {
+		auto &child = *children[child_idx];
+
+		if (sel.count != count) {
+			//! Some of the STRUCT rows are NULL entirely, we have to filter these rows out of the children
+			Vector new_child(child.GetType(), nullptr);
+			new_child.Dictionary(child, count, sel.non_null_selection, sel.count);
+			ToVariantSourceData csd(new_child, source.source_size);
+			if (!ConvertToVariant<WRITE_DATA, true>(csd, result, sel.count, &sel.new_selection, &sel.children_selection,
+			                                        false)) {
+				return false;
+			}
+		} else {
+			auto &csd = child_source_data[child_idx];
+			csd.source_sel = sel.non_null_selection;
+			if (!ConvertToVariant<WRITE_DATA, true>(csd, result, sel.count, &sel.new_selection, &sel.children_selection,
```

Comment:
Do note that this function doesn't work recursively, because `ConvertToVariant` will not call `ConvertStructToSparseVariant` for nested structs

Reply:
This is intended, it only does this for the first level of nesting.